## Проект по SQL

**Описание проекта:**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель проекта:**

   - Проанализировать базу данных;
   - Сформулировать ценностное предложение для нового продукта.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Исследуйте таблицы

In [3]:
query_books = '''SELECT * FROM books;'''
books = pd.io.sql.read_sql(query_books, con = engine)

query_authors = '''SELECT * FROM authors;'''
author = pd.io.sql.read_sql(query_authors, con = engine)

query_ratings = '''SELECT * FROM ratings'''
ratings = pd.io.sql.read_sql(query_ratings, con = engine)

query_reviews = '''SELECT * FROM reviews'''
reviews = pd.io.sql.read_sql(query_reviews, con=engine)

query_publishers = '''SELECT * FROM publishers'''
publishers = pd.io.sql.read_sql(query_publishers, con = engine)

In [4]:
datasets = [books, author, ratings, reviews, publishers]

for value in datasets:
    display(value.head(5))
    print(value.info())
    print('**'*50)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
****************************************************************************************************


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
****************************************************************************************************


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
****************************************************************************************************


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
****************************************************************************************************


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
****************************************************************************************************


### Задания
#### Посчитайте, сколько книг вышло после 1 января 2000 года

In [5]:
query_1 = '''
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '01.01.2000'
'''
task_1 = pd.io.sql.read_sql(query_1, con = engine)
print(f'После 1 января 2000 года вышло {task_1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


In [6]:
#Проверим в Pandas
books['publication_date'] = books['publication_date'].astype('datetime64')
books.query('publication_date >" 2000-01-01"')['book_id'].count()

819

**Получились одинаковые значения**

#### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [7]:
query_2 = '''
SELECT b.title,
       COUNT (DISTINCT re.review_id), ROUND(AVG(r.rating), 2) AS avg
FROM reviews re
RIGHT OUTER JOIN books AS b ON re.book_id=b.book_id
LEFT OUTER JOIN ratings AS r ON b.book_id = r.book_id
GROUP BY b.book_id
ORDER BY COUNT (DISTINCT re.review_id) DESC
'''
task_2 = pd.io.sql.read_sql(query_2, con = engine)
task_2

title  count   avg
0                              Twilight (Twilight  #1)      7  3.66
1                                  Water for Elephants      6  3.98
2                                     The Glass Castle      6  4.21
3    Harry Potter and the Prisoner of Azkaban (Harr...      6  4.41
4    The Curious Incident of the Dog in the Night-Time      6  4.08
..                                                 ...    ...   ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel      0  3.67
996                            The Natural Way to Draw      0  3.00
997   The Cat in the Hat and Other Dr. Seuss Favorites      0  5.00
998                          Essential Tales and Poems      0  4.00
999  Disney's Beauty and the Beast (A Little Golden...      0  4.00

[1000 rows x 3 columns]

In [8]:
#Проверим в Pandas
books_merge = books.merge(reviews, on='book_id',how='left')
books_merge = books_merge.merge(ratings, on ='book_id',how='left')

In [9]:
books_merge.groupby(['book_id','title']).agg({'review_id':'nunique','rating':'mean'}).sort_values(by='review_id',ascending=False)

review_id  \
book_id title                                                           
948     Twilight (Twilight  #1)                                     7   
963     Water for Elephants                                         6   
733     The Giver (The Giver  #1)                                   6   
750     The Hobbit  or There and Back Again                         6   
627     The Alchemist                                               6   
...                                                               ...   
808     The Natural Way to Draw                                     0   
191     Disney's Beauty and the Beast (A Little Golden ...          0   
221     Essential Tales and Poems                                   0   
387     Leonardo's Notebooks                                        0   
83      Anne Rice's The Vampire Lestat: A Graphic Novel             0   

                                                              rating  
book_id title                                                         
948     Twilight (Twilight  #1)                             3.662500  
963     Water for Elephants                                 3.977273  
733     The Giver (The Giver  #1)                           3.750000  
750     The Hobbit  or There and Back Again                 4.125000  
627     The Alchemist                                       3.789474  
...                                                              ...  
808     The Natural Way to Draw                             3.000000  
191     Disney's Beauty and the Beast (A Little Golden ...  4.000000  
221     Essential Tales and Poems                           4.000000  
387     Leonardo's Notebooks                                4.000000  
83      Anne Rice's The Vampire Lestat: A Graphic Novel     3.666667  

[1000 rows x 2 columns]

**Опять получились одинаковые значения**

#### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [10]:
query_3 = '''
SELECT *
FROM (
SELECT p.publisher, COUNT(DISTINCT b.book_id)
FROM books b
LEFT JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher) AS sub
ORDER BY count DESC
LIMIT 1
'''
task_3 = pd.io.sql.read_sql(query_3, con = engine)
print(f'Издательство, которое выпустило больше всего книг: {task_3.iloc[0,0]}. Количество книг: {task_3.iloc[0,1]}')

Издательство, которое выпустило больше всего книг: Penguin Books. Количество книг: 42


In [11]:
#Проверим в Pandas
books_merge_2 = books_merge.merge(publishers, on='publisher_id', how='left')
books_merge_2 = books_merge_2.query('num_pages > 50')
books_merge_2.groupby('publisher')['book_id'].nunique().sort_values(ascending=False)

publisher
Penguin Books                   42
Vintage                         31
Grand Central Publishing        25
Penguin Classics                24
Ballantine Books                19
                                ..
Devil's Due Publishing           1
Oxford University Press  USA     1
Pan Childrens                    1
Deodand                          1
Yearling Books                   1
Name: book_id, Length: 334, dtype: int64

**Получились одинаковые значения**

#### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [12]:
query_4 = '''
SELECT a.author, ROUND(AVG(r.rating),3)
FROM authors a 
JOIN books b ON b.author_id = a.author_id
JOIN ratings r ON b.book_id = r.book_id
WHERE b.book_id IN (SELECT book_id
FROM ratings
GROUP BY book_id
HAVING COUNT(rating_id) >=50
)
GROUP BY a.author
ORDER BY AVG(r.rating) DESC
LIMIT 1
'''
task_4 = pd.io.sql.read_sql(query_4, con = engine)
print(f'Автор с самой высокой средней оценкой книг {task_4.iloc[0,0]}, средняя оценка книг: {task_4.iloc[0,1]}')

Автор с самой высокой средней оценкой книг J.K. Rowling/Mary GrandPré, средняя оценка книг: 4.287


#### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [13]:
query_5 = '''
SELECT AVG(count)
FROM (
    SELECT COUNT(review_id)
    FROM reviews
    WHERE username IN 
        (SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 48
)
GROUP BY username) as sub
'''

task_5 = pd.io.sql.read_sql(query_5, con = engine)
task_5
print(f'Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: {task_5.iloc[0,0]}')

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24.0


**Вывод по проекту:**

Были рассчитаны 5 задач используя язык запросов SQL. В Pandas для точности выполнения задачи была проведена проверка.

Ответы на поставленные вопросы:
1. После 1 января 2000 года вышло 819 книг
2. Для 1000 книг посчиталаи количество обзоров и среднюю оценку
3. Издательство, которое выпустило больше всего книг: Penguin Books. Количество книг: 42
4. Определили авторов с самой высокой оценкой - J.K. Rowling/Mary GrandPré, средняя оценка книг: 4.287
5. Среднее количество обзоров от пользователей, которые поставили больше 48 оценок: 24.0